### ISSUE TITLE
Exception occurring in Foolscap CopyableSlicer.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2565]
|<img alt="kgi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kgi reported|
|-|-|
|Trac ID|trac#2565|
|Type|defect|
|Created|2007-04-11 18:15:25Z|

I've got an exception occurring in head-of-code SVN:

  'CopyableSlicer' object has no attribute 'streamable'

Which occurs here:

```
> /usr/lib/python2.4/site-packages/foolscap/banana.py(312)pushSlicer()
(Pdb) l
312  ->         itr = slicer.slice(topSlicer.streamable, self)
313             next = iter(itr).next
(Pdb) whatis topSlicer
<class 'foolscap.copyable.CopyableSlicer'>
(Pdb) dir(topSlicer)
['__class__', '__delattr__', '__dict__', '__doc__', '__getattribute__', '__hash__',
'__implemented__', '__init__', '__module__', '__new__', '__providedBy__',
'__provides__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__',
'__str__', '__weakref__', 'childAborted', 'describe', 'obj', 'opentype',
'parent', 'registerReference', 'sendOpen', 'slice', 'sliceBody',
'slicerForObject', 'slices', 'trackReferences']
```

So the error happened when trying to access topSlicer.streamable, and indeed, 
dir() shows there is no such attribute on the object, which is of type 
CopyableSlicer. Now, CopyableSlicer is defined like this:

```
class CopyableSlicer(slicer.BaseSlicer):
    """I handle ICopyable objects (things which are copied by value)."""
    def slice(self, streamable, banana):
        yield 'copyable'
        copytype = self.obj.getTypeToCopy()
        assert isinstance(copytype, str)
        yield copytype
        state = self.obj.getStateToCopy()
        for k,v in state.iteritems():
            yield k
            yield v
    def describe(self):
        return "<%s>" % self.obj.getTypeToCopy()
registerAdapter(CopyableSlicer, ICopyable, tokens.ISlicer)
```

So it overrides slicer.BaseSlicer.slice(). Now, in slicer.BaseSlicer, we see:

```
class BaseSlicer:
    __metaclass__ = SlicerClass
    implements(tokens.ISlicer)
    ....
    def slice(self, streamable, banana):
        # this is what makes us ISlicer
        self.streamable = streamable
        assert self.opentype
        for o in self.opentype:
            yield o
        for t in self.sliceBody(streamable, banana):
            yield t
```

Note the line following the comment, "# this is what makes us ISlicer" sets 
a 'streamable' attribute on the object in the base class, but the subclass 
doesn't preserve this behaviour.

<details><summary>Searchable metadata</summary>

```
trac-id__2565 2565
type__defect defect
reporter__kgi kgi
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__foolscap foolscap
keywords__ 
time__1176315325000000 1176315325000000
changetime__1179209967000000 1179209967000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_1"></a>|@warner set status to `assigned`|
|-|-|

good catch. I think I've fixed this in trunk, but we still need a unit test coverage for this. The specific problem didn't show up until the serializer encountered a Copyable while in "streaming" mode.. and to be honest I haven't ever used streaming mode, so I'm not sure what other problems might still be lurking.

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_2"></a>|@warner set status to `closed`|
|-|-|

this fix made it into the 0.1.3 release, so I'll close out this ticket.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
